File: plugin/api/index_management/indices.go
// Copyright (C) INFINI Labs & INFINI LIMITED.
//
// The INFINI Console is offered under the GNU Affero General Public License v3.0
// and as commercial software.
//
// For commercial licensing, contact us at:
//   - Website: infinilabs.com
//   - Email: hello@infini.ltd
//
// Open Source licensed under AGPL V3:
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program. If not, see <http://www.gnu.org/licenses/>.

package index_management

import (
	log "github.com/cihub/seelog"
	"infini.sh/console/common"
	"infini.sh/console/core/security"
	"infini.sh/console/model"
	"infini.sh/console/service"
	httprouter "infini.sh/framework/core/api/router"
	"infini.sh/framework/core/elastic"
	"infini.sh/framework/core/radix"
	"infini.sh/framework/core/util"
	"net/http"
)

func (handler APIHandler) HandleGetMappingsAction(w http.ResponseWriter, req *http.Request, ps httprouter.Params) {
	targetClusterID := ps.ByName("id")
	client := elastic.GetClient(targetClusterID)
	indexName := ps.ByName("index")
	resBody := newResponseBody()
	var copyAll = false
	if indexName == "*" {
		indexName = ""
		copyAll = true
	}
	_, _, idxs, err := client.GetMapping(copyAll, indexName)
	if err != nil {
		log.Error(err)
		resBody["error"] = err.Error()
		handler.WriteJSON(w, resBody, http.StatusInternalServerError)
		return
	}
	//if copyAll {
	//	for key, _ := range *idxs {
	//		if strings.HasPrefix(key, ".") || strings.HasPrefix(key, "infini-") {
	//			delete(*idxs, key)
	//		}
	//	}
	//}

	handler.WriteJSON(w, idxs, http.StatusOK)
}

func (handler APIHandler) HandleCatIndicesAction(w http.ResponseWriter, req *http.Request, ps httprouter.Params) {
	targetClusterID := ps.ByName("id")
	user, auditLogErr := security.FromUserContext(req.Context())
	if auditLogErr == nil && handler.GetHeader(req, "Referer", "") != "" {
		auditLog, _ := model.NewAuditLogBuilderWithDefault().WithOperator(user.Username).
			WithLogTypeAccess().WithResourceTypeClusterManagement().
			WithEventName("get indices").WithEventSourceIP(common.GetClientIP(req)).
			WithResourceName(targetClusterID).WithOperationTypeAccess().WithEventRecord("").Build()
		_ = service.LogAuditLog(auditLog)
	}
	client := elastic.GetClient(targetClusterID)
	//filter indices
	allowedIndices, hasAllPrivilege := handler.GetAllowedIndices(req, targetClusterID)
	if !hasAllPrivilege && len(allowedIndices) == 0 {
		handler.WriteJSON(w, []interface{}{}, http.StatusOK)
		return
	}
	catIndices, err := client.GetIndices("")
	resBody := util.MapStr{}
	if err != nil {
		log.Error(err)
		resBody["error"] = err.Error()
		handler.WriteJSON(w, resBody, http.StatusInternalServerError)
		return
	}
	if !hasAllPrivilege {
		filterIndices := map[string]elastic.IndexInfo{}
		pattern := radix.Compile(allowedIndices...)
		for indexName, indexInfo := range *catIndices {
			if pattern.Match(indexName) {
				filterIndices[indexName] = indexInfo
			}
		}
		catIndices = &filterIndices
	}
	handler.WriteJSON(w, catIndices, http.StatusOK)
}

func (handler APIHandler) HandleGetSettingsAction(w http.ResponseWriter, req *http.Request, ps httprouter.Params) {
	targetClusterID := ps.ByName("id")
	client := elastic.GetClient(targetClusterID)
	indexName := ps.ByName("index")
	resBody := newResponseBody()
	indexes, err := client.GetIndexSettings(indexName)
	if err != nil {
		log.Error(err)
		resBody["error"] = err.Error()
		handler.WriteJSON(w, resBody, http.StatusInternalServerError)
		return
	}
	handler.WriteJSON(w, indexes, http.StatusOK)
}

func (handler APIHandler) HandleUpdateSettingsAction(w http.ResponseWriter, req *http.Request, ps httprouter.Params) {
	targetClusterID := ps.ByName("id")
	client := elastic.GetClient(targetClusterID)
	indexName := ps.ByName("index")
	settings := map[string]interface{}{}
	resBody := newResponseBody()
	err := handler.DecodeJSON(req, &settings)
	if err != nil {
		log.Error(err)
		resBody["error"] = err.Error()
		handler.WriteJSON(w, resBody, http.StatusInternalServerError)
		return
	}
	err = client.UpdateIndexSettings(indexName, settings)
	if err != nil {
		log.Error(err)
		resBody["error"] = err.Error()
		handler.WriteJSON(w, resBody, http.StatusInternalServerError)
		return
	}
	resBody["result"] = "updated"
	handler.WriteJSON(w, resBody, http.StatusCreated)
}

func (handler APIHandler) HandleDeleteIndexAction(w http.ResponseWriter, req *http.Request, ps httprouter.Params) {
	targetClusterID := ps.ByName("id")
	client := elastic.GetClient(targetClusterID)
	indexName := ps.ByName("index")
	resBody := newResponseBody()
	err := client.DeleteIndex(indexName)
	if err != nil {
		log.Error(err)
		resBody["error"] = err.Error()
		handler.WriteJSON(w, resBody, http.StatusInternalServerError)
		return
	}
	resBody["result"] = "deleted"
	handler.WriteJSON(w, resBody, http.StatusOK)
}

func (handler APIHandler) HandleCreateIndexAction(w http.ResponseWriter, req *http.Request, ps httprouter.Params) {

	targetClusterID := ps.ByName("id")
	client := elastic.GetClient(targetClusterID)
	indexName := ps.ByName("index")
	claims, auditLogErr := security.ValidateLogin(req.Header.Get("Authorization"))
	if auditLogErr == nil && handler.GetHeader(req, "Referer", "") != "" {
		auditLog, _ := model.NewAuditLogBuilderWithDefault().WithOperator(claims.Username).
			WithLogTypeOperation().WithResourceTypeClusterManagement().
			WithEventName("create index").WithEventSourceIP(common.GetClientIP(req)).
			WithResourceName(targetClusterID).WithOperationTypeNew().WithEventRecord(indexName).Build()
		auditLogErr = service.LogAuditLog(auditLog)
	}
	resBody := newResponseBody()
	config := map[string]interface{}{}
	err := handler.DecodeJSON(req, &config)
	if err != nil {
		log.Error(err)
		resBody["error"] = err.Error()
		handler.WriteJSON(w, resBody, http.StatusInternalServerError)
		return
	}
	err = client.CreateIndex(indexName, config)
	if err != nil {
		log.Error(err)
		resBody["error"] = err.Error()
		handler.WriteJSON(w, resBody, http.StatusInternalServerError)
		return
	}
	resBody["result"] = "created"
	handler.WriteJSON(w, resBody, http.StatusCreated)
}

func (handler APIHandler) HandleGetIndexAction(w http.ResponseWriter, req *http.Request, ps httprouter.Params) {
	targetClusterID := ps.ByName("id")
	client := elastic.GetClient(targetClusterID)
	indexName := ps.ByName("index")
	indexRes, err := client.GetIndex(indexName)
	if err != nil {
		log.Error(err)
		handler.WriteError(w, err.Error(), http.StatusInternalServerError)
		return
	}
	handler.WriteJSON(w, indexRes, http.StatusOK)

}

File: web/mock/datamanagement/indices.js
// Copyright (C) INFINI Labs & INFINI LIMITED.
//
// The INFINI Console is offered under the GNU Affero General Public License v3.0
// and as commercial software.
//
// For commercial licensing, contact us at:
//   - Website: infinilabs.com
//   - Email: hello@infini.ltd
//
// Open Source licensed under AGPL V3:
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program. If not, see <http://www.gnu.org/licenses/>.

const data = {
  payload: {
    blogs: {
      id: "3YOmOx_BSDqF-6EqAzBLVg",
      index: "blogs",
      status: "open",
      health: "green",
      shards: 1,
      replicas: 1,
      docs_count: 3,
    },
    dict: {
      id: "F6vp1k_XRn-FmXoHOsBl2Q",
      index: "dict",
      status: "open",
      health: "green",
      shards: 1,
      replicas: 1,
      docs_count: 8,
      docs_deleted: 1,
    },
    gateway_requests: {
      id: "C0j0942KR6muJMHyagG2AQ",
      index: "gateway_requests",
      status: "open",
      health: "green",
      shards: 1,
      replicas: 1,
      docs_count: 17961,
    },
    "metricbeat-7.10.0-2020.11.24-000001": {
      id: "j5hJlvknQliWmVvdLVSH7w",
      index: "metricbeat-7.10.0-2020.11.24-000001",
      status: "open",
      health: "green",
      shards: 1,
      replicas: 1,
      docs_count: 257315,
    },
    "metricbeat-7.10.0-2020.12.24-000002": {
      id: "Q3f7CTSfScCLTUCBKURzCw",
      index: "metricbeat-7.10.0-2020.12.24-000002",
      status: "open",
      health: "green",
      shards: 1,
      replicas: 1,
    },
    mock_log: {
      id: "NI7ntH_YRHapvjqB30LFQA",
      index: "mock_log",
      status: "open",
      health: "green",
      shards: 1,
      replicas: 1,
      docs_count: 45806,
    },
    mock_log1: {
      id: "FEuAjVAYTCG5FJSVqjB4cA",
      index: "mock_log1",
      status: "open",
      health: "green",
      shards: 1,
      replicas: 1,
      docs_count: 45806,
    },
    reindex: {
      id: "o7x8G6csQbyAMrWYqZOojA",
      index: "reindex",
      status: "open",
      health: "green",
      shards: 1,
      replicas: 1,
    },
    "test-custom": {
      id: "17I4JLDWRdGrBL15sL3qIA",
      index: "test-custom",
      status: "open",
      health: "green",
      shards: 1,
      replicas: 1,
      docs_count: 12,
      docs_deleted: 1,
    },
    "test-custom1": {
      id: "SEpAkmImQsGD0zVCHMtteQ",
      index: "test-custom1",
      status: "open",
      health: "green",
      shards: 1,
      replicas: 1,
      docs_count: 4,
    },
    "test-custom8": {
      id: "PyHOJ_ytQxCzeamGBhs1_Q",
      index: "test-custom8",
      status: "open",
      health: "green",
      shards: 1,
      replicas: 1,
      docs_count: 12,
    },
  },
  status: true,
};

const mappings = {
  payload: {
    blogs: {
      mappings: {
        properties: {
          created_at: {
            type: "date",
          },
          test_field: {
            fields: {
              keyword: {
                ignore_above: 256,
                type: "keyword",
              },
            },
            type: "text",
          },
          test_field2: {
            fields: {
              keyword: {
                ignore_above: 256,
                type: "keyword",
              },
            },
            type: "text",
          },
          title: {
            fields: {
              keyword: {
                ignore_above: 256,
                type: "keyword",
              },
            },
            type: "text",
          },
          view_count: {
            type: "long",
          },
        },
      },
    },
    dict: {
      mappings: {
        properties: {
          content: {
            type: "binary",
          },
          created_at: {
            type: "date",
          },
          id: {
            fields: {
              keyword: {
                ignore_above: 256,
                type: "keyword",
              },
            },
            type: "text",
          },
          name: {
            type: "text",
          },
          tags: {
            type: "text",
          },
          updated_at: {
            type: "date",
          },
        },
      },
    },
    gateway_requests: {
      mappings: {
        dynamic_templates: [
          {
            strings: {
              mapping: {
                ignore_above: 256,
                type: "keyword",
              },
              match_mapping_type: "string",
            },
          },
        ],
        properties: {
          "@timestamp": {
            type: "date",
          },
          conn_time: {
            type: "date",
          },
          flow: {
            properties: {
              from: {
                ignore_above: 256,
                type: "keyword",
              },
              relay: {
                ignore_above: 256,
                type: "keyword",
              },
              to: {
                ignore_above: 256,
                type: "keyword",
              },
            },
          },
          id: {
            type: "long",
          },
          local_ip: {
            ignore_above: 256,
            type: "keyword",
          },
          remote_ip: {
            ignore_above: 256,
            type: "keyword",
          },
          request: {
            properties: {
              body: {
                ignore_above: 256,
                type: "keyword",
              },
              body_length: {
                type: "long",
              },
              header: {
                properties: {
                  accept: {
                    ignore_above: 256,
                    type: "keyword",
                  },
                  "accept-encoding": {
                    ignore_above: 256,
                    type: "keyword",
                  },
                  "accept-language": {
                    ignore_above: 256,
                    type: "keyword",
                  },
                  "cache-control": {
                    ignore_above: 256,
                    type: "keyword",
                  },
                  connection: {
                    ignore_above: 256,
                    type: "keyword",
                  },
                  "content-length": {
                    ignore_above: 256,
                    type: "keyword",
                  },
                  "content-type": {
                    ignore_above: 256,
                    type: "keyword",
                  },
                  content_type: {
                    ignore_above: 256,
                    type: "keyword",
                  },
                  cookie: {
                    ignore_above: 256,
                    type: "keyword",
                  },
                  host: {
                    ignore_above: 256,
                    type: "keyword",
                  },
                  "if-none-match": {
                    ignore_above: 256,
                    type: "keyword",
                  },
                  pragma: {
                    ignore_above: 256,
                    type: "keyword",
                  },
                  referer: {
                    ignore_above: 256,
                    type: "keyword",
                  },
                  "sec-fetch-dest": {
                    ignore_above: 256,
                    type: "keyword",
                  },
                  "sec-fetch-mode": {
                    ignore_above: 256,
                    type: "keyword",
                  },
                  "sec-fetch-site": {
                    ignore_above: 256,
                    type: "keyword",
                  },
                  "sec-fetch-user": {
                    ignore_above: 256,
                    type: "keyword",
                  },
                  "upgrade-insecure-requests": {
                    ignore_above: 256,
                    type: "keyword",
                  },
                  "user-agent": {
                    ignore_above: 256,
                    type: "keyword",
                  },
                },
              },
              host: {
                ignore_above: 256,
                type: "keyword",
              },
              local_addr: {
                ignore_above: 256,
                type: "keyword",
              },
              method: {
                ignore_above: 256,
                type: "keyword",
              },
              path: {
                ignore_above: 256,
                type: "keyword",
              },
              query_args: {
                type: "object",
              },
              remote_addr: {
                ignore_above: 256,
                type: "keyword",
              },
              started: {
                type: "date",
              },
              uri: {
                ignore_above: 256,
                type: "keyword",
              },
            },
          },
          response: {
            properties: {
              body: {
                ignore_above: 256,
                type: "keyword",
              },
              body_length: {
                type: "long",
              },
              cached: {
                type: "boolean",
              },
              elapsed: {
                type: "float",
              },
              header: {
                properties: {
                  "content-encoding": {
                    ignore_above: 256,
                    type: "keyword",
                  },
                  "content-length": {
                    ignore_above: 256,
                    type: "keyword",
                  },
                  "content-type": {
                    ignore_above: 256,
                    type: "keyword",
                  },
                  "infini-cache": {
                    ignore_above: 256,
                    type: "keyword",
                  },
                  server: {
                    ignore_above: 256,
                    type: "keyword",
                  },
                },
              },
              local_addr: {
                ignore_above: 256,
                type: "keyword",
              },
              remote_addr: {
                ignore_above: 256,
                type: "keyword",
              },
              status_code: {
                type: "long",
              },
            },
          },
          tls: {
            type: "boolean",
          },
        },
      },
    },
    mock_log: {
      mappings: {
        properties: {
          method: {
            type: "keyword",
          },
          msg: {
            fields: {
              keyword: {
                ignore_above: 256,
                type: "keyword",
              },
            },
            type: "text",
          },
          timestamp: {
            type: "date",
          },
        },
      },
    },
    mock_log1: {
      mappings: {
        properties: {
          method: {
            fields: {
              keyword: {
                ignore_above: 256,
                type: "keyword",
              },
            },
            type: "text",
          },
          msg: {
            fields: {
              keyword: {
                ignore_above: 256,
                type: "keyword",
              },
            },
            type: "text",
          },
          timestamp: {
            type: "long",
          },
        },
      },
    },
    reindex: {
      mappings: {
        properties: {
          created_at: {
            type: "date",
          },
          desc: {
            type: "text",
          },
          dest: {
            type: "object",
          },
          name: {
            type: "text",
          },
          source: {
            type: "object",
          },
          status: {
            type: "keyword",
          },
          task_id: {
            type: "keyword",
          },
        },
      },
    },
    "test-custom": {
      mappings: {
        properties: {
          address: {
            fields: {
              keyword: {
                ignore_above: 256,
                type: "keyword",
              },
            },
            type: "text",
          },
          age: {
            type: "long",
          },
          created_at: {
            type: "date",
          },
          email: {
            fields: {
              keyword: {
                ignore_above: 256,
                type: "keyword",
              },
            },
            type: "text",
          },
          hobbies: {
            fields: {
              keyword: {
                ignore_above: 256,
                type: "keyword",
              },
            },
            type: "text",
          },
          id: {
            fields: {
              keyword: {
                ignore_above: 256,
                type: "keyword",
              },
            },
            type: "text",
          },
          name: {
            fields: {
              keyword: {
                ignore_above: 256,
                type: "keyword",
              },
            },
            type: "text",
          },
        },
      },
    },
    "test-custom1": {
      mappings: {
        properties: {
          address: {
            fields: {
              keyword: {
                ignore_above: 256,
                type: "keyword",
              },
            },
            type: "text",
          },
          created_at: {
            type: "date",
          },
          email: {
            fields: {
              keyword: {
                ignore_above: 256,
                type: "keyword",
              },
            },
            type: "text",
          },
          hobbies: {
            fields: {
              keyword: {
                ignore_above: 256,
                type: "keyword",
              },
            },
            type: "text",
          },
          id: {
            fields: {
              keyword: {
                ignore_above: 256,
                type: "keyword",
              },
            },
            type: "text",
          },
          name: {
            fields: {
              keyword: {
                ignore_above: 256,
                type: "keyword",
              },
            },
            type: "text",
          },
        },
      },
    },
    "test-custom8": {
      mappings: {
        properties: {
          address: {
            fields: {
              keyword: {
                ignore_above: 256,
                type: "keyword",
              },
            },
            type: "text",
          },
          age: {
            type: "long",
          },
          created_at: {
            type: "date",
          },
          email: {
            fields: {
              keyword: {
                ignore_above: 256,
                type: "keyword",
              },
            },
            type: "text",
          },
          hobbies: {
            fields: {
              keyword: {
                ignore_above: 256,
                type: "keyword",
              },
            },
            type: "text",
          },
          id: {
            fields: {
              keyword: {
                ignore_above: 256,
                type: "keyword",
              },
            },
            type: "text",
          },
          name: {
            fields: {
              keyword: {
                ignore_above: 256,
                type: "keyword",
              },
            },
            type: "text",
          },
        },
      },
    },
  },
  status: true,
};

export default {
  // 'get /_search-center/_cat/indices': function(req, res){
  //     res.send(data)
  // },
  // 'get /_search-center/index/:index/_mappings': function(req, res){
  //     res.send(mappings)
  // }

  "GET /elasticsearch/:cluster_id/_cat/indices": function(req, res) {
    res.send({
      ".apm-agent-configuration": {
        id: "DXT39lbQQFCY9p4_XTUTFQ",
        index: ".apm-agent-configuration",
        status: "open",
        health: "green",
        shards: 1,
        replicas: 1,
        store_size: "416b",
        pri_store_size: "208b",
      },
      ".apm-custom-link": {
        id: "bRw9BgQTQSyIxYqeCzf6jg",
        index: ".apm-custom-link",
        status: "open",
        health: "green",
        shards: 1,
        replicas: 1,
        store_size: "416b",
        pri_store_size: "208b",
      },
      ".kibana-event-log-7.9.3-000001": {
        id: "kjmVS-oTT2q7qEbYNY8nIQ",
        index: ".kibana-event-log-7.9.3-000001",
        status: "open",
        health: "green",
        shards: 1,
        replicas: 1,
        docs_count: 1,
        segments_count: 2,
        store_size: "11.1kb",
        pri_store_size: "5.5kb",
      },
      ".kibana_1": {
        id: "tDgyKEadQMGoTlVcEY2YPw",
        index: ".kibana_1",
        status: "open",
        health: "green",
        shards: 1,
        replicas: 1,
        docs_count: 14,
        segments_count: 12,
        store_size: "20.8mb",
        pri_store_size: "10.4mb",
      },
      ".kibana_task_manager_1": {
        id: "h_3tImGORHCplX3-UMvOpg",
        index: ".kibana_task_manager_1",
        status: "open",
        health: "green",
        shards: 1,
        replicas: 1,
        docs_count: 6,
        docs_deleted: 60,
        segments_count: 16,
        store_size: "197.3kb",
        pri_store_size: "109.9kb",
      },
      "filebeat-7.8.0-2022.03.08-000001": {
        id: "KWBOoelORmiJmkBce_MdXg",
        index: "filebeat-7.8.0-2022.03.08-000001",
        status: "open",
        health: "green",
        shards: 1,
        replicas: 1,
        docs_count: 21409646,
        segments_count: 63,
        store_size: "10gb",
        pri_store_size: "5gb",
      },
      "gateway_requests": {
        id: "vTVZSxwIR7qyYEYkw7Q90w",
        index: "gateway_requests",
        status: "open",
        health: "green",
        shards: 1,
        replicas: 1,
        docs_count: 206582,
        segments_count: 66,
        store_size: "9.5gb",
        pri_store_size: "5.4gb",
      },
    });
  },
};
